### PR TITLE
nixos/systemd: add an option for ExecCondition

### PR DESCRIPTION
**Pull request by danth**
###### Motivation for this change

The other `Exec*` options in SystemD already have NixOS equivalents, so it makes sense to add this one for consistency.

A description of what `ExecCondition` does can be found [here](https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecCondition=).

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/systemd.nix [hunk 1]
```diff
@@ -294,6 +294,10 @@ let
           { serviceConfig.ExecStopPost =
               makeJobScript "${name}-post-stop" config.postStop;
           })
+        (mkIf (config.condition != "")
+          { serviceConfig.ExecCondition =
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/systemd.nix:298]: Please sort this alphabetical.

File: nixos/modules/system/boot/systemd-unit-options.nix [hunk 2]
```diff
@@ -321,6 +321,15 @@ in rec {
       '';
     };
 
+    condition = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Shell commands executed to determine whether the service
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/systemd-unit-options.nix:328]: This can be written without the line break.

**Comment by Ma27** [comment]
Does this have any benefits over `serviceConfig.ExecCondition`?

The thing is, `script` is e.g. only a historical artifact from a past where NixOS used upstart. Also, it's IIMHO not a nice thing to use these options since most things are one-liners (that can be split over multiple lines in `.service`-files) that are now written to its own store-path which introduces another level of indirection. This means that when I e.g. run `systemctl cat foo.service`, I only see a store-path for `Exec*`-settings rather than what's actually happening.

In all other cases I don't see what's wrong with `${pkgs.writeShellScript "foo" "bar"}`.

**Comment by SuperSandro2000** [comment]
> The thing is, script is e.g. only a historical artifact from a past where NixOS used upstart.

Off topic question but should we remove the other execs and redirect them to serviceConfig?

**Comment by danth** [comment]
> Does this have any benefits over `serviceConfig.ExecCondition`?

This and the other script options are passed through `makeJobScript`:

https://github.com/NixOS/nixpkgs/blob/a97f20e5d822a373153dca35058462dc00a0cf5b/nixos/modules/system/boot/systemd.nix#L209-L226

Scripts are written into a sub-directory to avoid cluttering the log with lengthy derivation names.

To replicate this behaviour using only `serviceConfig.ExecCondition`, you would need to write a complex expression like `${"${pkgs.writeShellScriptBin "foo" "bar"}/bin/foo"}`. This is especially difficult to read when bar is a multi-line string.

> Also, it's IIMHO not a nice thing to use these options since most things are one-liners

Would it be possible for `makeJobScript` to return one-liners verbatim while still writing a file for longer scripts?

**Pull request closed without merging**